// fetch-mock-cache/lib/fetch-cache.js
import _debug from "debug";
import { deserializeBody, serializeBody } from "./body.js";
import { DEFAULT_REDACTED_HEADERS, deserializeHeaders, redactHeaders as redactHeadersHelper, serializeHeaders, } from "./headers.js";
import { DEFAULT_REDACTED_SEARCH_PARAMS, redactSearchParams as redactUrlSearchParams, } from "./url.js";
const debug = _debug("fetch-mock-cache:core");
const origFetch = fetch;
const VALID_FETCH_CACHE_MODES = ["auto", "replay", "record", "off"];
const FETCH_CACHE_MODE_DEFAULTS = {
    auto: { readCache: true, writeCache: true, throwOnMiss: false },
    replay: { readCache: true, writeCache: false, throwOnMiss: true },
    record: { readCache: false, writeCache: true, throwOnMiss: false },
    off: { readCache: false, writeCache: false, throwOnMiss: false },
};
/**
 * Creates a new caching fetch implementation.  Generally not used directly,
 * instead use the same named function provided by the various run-time entry
 * points.
 */
export default function createCachingMock({ Store, fetch, runtime, id, mode, readCache, writeCache, redactHeaders, redactRequestHeaders, redactResponseHeaders, redactSearchParams, }) {
    if (!Store) {
        throw new Error("No `Store` option was provided, but is required.  See docs.");
    }
    if (!fetch)
        fetch = origFetch;
    const resolveHeaderRedactList = (specific) => specific === false
        ? null
        : (specific ??
            (redactHeaders === false
                ? null
                : (redactHeaders ?? DEFAULT_REDACTED_HEADERS)));
    const requestHeaderRedactList = resolveHeaderRedactList(redactRequestHeaders);
    const responseHeaderRedactList = resolveHeaderRedactList(redactResponseHeaders);
    const paramRedactList = redactSearchParams === false
        ? null
        : (redactSearchParams ?? DEFAULT_REDACTED_SEARCH_PARAMS);
    const serializeAndRedactHeaders = (headers, redactList) => {
        const serialized = serializeHeaders(headers);
        return redactList
            ? redactHeadersHelper(serialized, redactList)
            : serialized;
    };
    // Init with options if passed as [ Store, { /* ... */ } ]
    const store = Array.isArray(Store)
        ? new Store[0]({ ...Store[1], runtime })
        : new Store({ runtime });
    const defaultOptions = mergeFetchCacheOptions({
        id,
        mode,
        readCache,
        writeCache,
    });
    const fetchCache = Object.assign(async function cachingMockImplementation(urlOrRequest, requestInit) {
        if (!urlOrRequest)
            throw new Error("urlOrRequest is undefined");
        const onceOptions = Array.isArray(fetchCache._options)
            ? fetchCache._options.shift()
            : fetchCache._options;
        const modeResolution = resolveFetchCacheMode(onceOptions, fetchCache.options, runtime);
        const modeDefaults = FETCH_CACHE_MODE_DEFAULTS[modeResolution.mode];
        const options = mergeFetchCacheOptions(fetchCache.options, onceOptions);
        const writeCache = resolveOptionWithModeDefaults("writeCache", onceOptions, fetchCache.options, modeResolution.sourcePriority, modeDefaults.writeCache);
        const fetchRequest = typeof urlOrRequest === "string" || urlOrRequest instanceof URL
            ? new Request(urlOrRequest, requestInit)
            : urlOrRequest;
        const url = fetchRequest.url;
        const cacheUrl = paramRedactList
            ? redactUrlSearchParams(url, paramRedactList)
            : url;
        const clonedRequest = fetchRequest.clone();
        const cacheContentRequest = {
            url: cacheUrl,
            method: fetchRequest.method,
            ...(clonedRequest.body && (await serializeBody(clonedRequest))),
            ...(Array.from(fetchRequest.headers.keys()).length > 0 && {
                // Not really necessary as set-cookie never appears in the REQUEST headers.
                headers: serializeAndRedactHeaders(fetchRequest.headers, requestHeaderRedactList),
            }),
        };
        const readCache = await resolveReadCacheOption(resolveOptionWithModeDefaults("readCache", onceOptions, fetchCache.options, modeResolution.sourcePriority, modeDefaults.readCache), cacheContentRequest, options);
        const existingContent = readCache && (await store.fetchContent(cacheContentRequest, options));
        if (existingContent) {
            debug("Using cached copy of %o", cacheUrl);
            const headers = deserializeHeaders(existingContent.response.headers);
            headers.set("X-FMC-Cache", "HIT");
            return new Response(await deserializeBody(existingContent.response), {
                status: existingContent.response.status,
                statusText: existingContent.response.statusText,
                headers,
            });
        }
        if (modeDefaults.throwOnMiss && readCache) {
            throw await replayMissError(store, cacheContentRequest, options);
        }
        debug("Fetching %o", cacheUrl);
        const response = typeof urlOrRequest === "string" || urlOrRequest instanceof URL
            ? await fetch(urlOrRequest, requestInit)
            : await fetch(fetchRequest);
        const newContent = {
            request: cacheContentRequest,
            response: {
                ok: response.ok,
                status: response.status,
                statusText: response.statusText,
                headers: serializeAndRedactHeaders(response.headers, responseHeaderRedactList),
                ...(await serializeBody(response)),
            },
        };
        const resolvedWriteCache = typeof writeCache === "function"
            ? writeCache(newContent, options)
            : writeCache;
        if (resolvedWriteCache instanceof Promise) {
            resolvedWriteCache
                .then(async (shouldWrite) => {
                if (shouldWrite) {
                    await store.storeContent(newContent, options);
                }
            })
                .catch((error) => {
                console.error("Error occurred while deciding to cache response: %o", error);
            });
        }
        else if (resolvedWriteCache)
            await store.storeContent(newContent, options);
        const headers = new Headers(response.headers);
        headers.set("X-FMC-Cache", "MISS");
        return new Response(await deserializeBody(newContent.response), {
            status: response.status,
            statusText: response.statusText,
            headers,
        });
    }, {
        runtime,
        options: defaultOptions,
        _store: store,
        _options: [], // TODO
        once(options) {
            if (!Array.isArray(this._options))
                this._options = [];
            this._options.push(options);
        },
        _once(options) {
            console.warn("_fetchCache._once() is deprecated, use _fetchCache.once()");
            return this.once(options);
        },
    });
    return fetchCache;
}
function mergeFetchCacheOptions(...sources) {
    const options = {};
    for (const source of sources) {
        if (!source)
            continue;
        if (source.id !== undefined)
            options.id = source.id;
        if (source.mode !== undefined)
            options.mode = source.mode;
        if (source.readCache !== undefined)
            options.readCache = source.readCache;
        if (source.writeCache !== undefined) {
            options.writeCache = source.writeCache;
        }
    }
    return options;
}
function resolveFetchCacheMode(onceOptions, globalOptions, runtime) {
    if (onceOptions?.mode !== undefined) {
        return {
            mode: parseFetchCacheMode(onceOptions.mode, "once()"),
            sourcePriority: 2,
        };
    }
    if (globalOptions?.mode !== undefined) {
        return {
            mode: parseFetchCacheMode(globalOptions.mode, "fetchCache.options"),
            sourcePriority: 1,
        };
    }
    const envMode = runtime.env.FMC_CACHE_MODE;
    if (envMode !== undefined) {
        return {
            mode: parseFetchCacheMode(envMode, "FMC_CACHE_MODE"),
            sourcePriority: 0,
        };
    }
    return { mode: "auto", sourcePriority: 0 };
}
function parseFetchCacheMode(value, source) {
    const mode = value.trim().toLowerCase();
    if (isFetchCacheMode(mode))
        return mode;
    throw new Error(`fetch-mock-cache: invalid cache mode ${JSON.stringify(value)} from ${source}. Valid modes: ${VALID_FETCH_CACHE_MODES.join(", ")}.`);
}
function isFetchCacheMode(mode) {
    return VALID_FETCH_CACHE_MODES.includes(mode);
}
function resolveOptionWithModeDefaults(optionName, onceOptions, globalOptions, modeSourcePriority, defaultValue) {
    if (onceOptions?.[optionName] !== undefined && modeSourcePriority <= 2) {
        return onceOptions[optionName];
    }
    if (globalOptions?.[optionName] !== undefined && modeSourcePriority <= 1) {
        return globalOptions[optionName];
    }
    return defaultValue;
}
async function resolveReadCacheOption(readCache, request, options) {
    if (typeof readCache === "function") {
        return await readCache(request, options);
    }
    return await readCache;
}
async function replayMissError(store, request, options) {
    const storeName = store.constructor.name || "FMCStore";
    const storeLocation = "_location" in store && typeof store._location === "string"
        ? ` (location: ${store._location})`
        : "";
    const method = request.method || "GET";
    const cacheId = await store.idFromRequest(request, options);
    return new Error([
        `fetch-mock-cache: cache miss in replay mode for ${method} ${request.url}`,
        `Store: ${storeName}${storeLocation}`,
        `Computed cache id: ${cacheId}`,
        'To record this fixture, set FMC_CACHE_MODE=record or createFetchCache({ mode: "record" }).',
    ].join("\n"));
}
