// fetch-mock-cache/lib/runtimes/node.js
/**
 * Entry point for using fetch-mock-cache with the Node runtime.
 * @module
 */
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import _createFetchCache from "../fetch-cache.js";
/**
 * Runtime adapter that provides Node.js environment, hashing, filesystem,
 * path, and cwd operations to stores.
 */
export const runtime = {
    name: "node",
    get env() {
        return process.env;
    },
    async sha256(input, length) {
        const utf8 = new TextEncoder().encode(input);
        const hashBuffer = await crypto.subtle.digest("SHA-256", utf8);
        const hashArray = Array.from(new Uint8Array(hashBuffer));
        const hashHex = hashArray
            .map((bytes) => bytes.toString(16).padStart(2, "0"))
            .join("");
        return length ? hashHex.substring(0, length) : hashHex;
    },
    fs: {
        readFile: async (path) => fs.promises.readFile(path, "utf8"),
        writeFile: fs.promises.writeFile,
        mkdir: fs.promises.mkdir,
    },
    path: {
        join: path.join,
    },
    cwd: process.cwd,
};
/**
 * Creates a cached `fetch` implementation using the Node.js runtime adapter.
 *
 * @example
 * ```ts
 * import createFetchCache from "fetch-mock-cache/runtimes/node.js"
 * import Store from "fetch-mock-cache/stores/memory";
 * const fetchCache = createFetchCache({ Store });
 * ```
 */
export default function createFetchCache(options = {}) {
    return _createFetchCache({ ...options, runtime });
}
